### ISSUE TITLE
CVE-2018-3615:KO shows microcode not up to date, when, in fact, it is

### ISSUE THREAD

**Issue by eris23** [speed47/spectre-meltdown-checker#296]
I've been exploring this problem at https://answers.launchpad.net/ubuntu/+source/intel-microcode/+question/681827.  It doesn't seem to be a problem on the Ubuntu side.  Is this being falsely flagged?

**Comment by hotaru2k3** [comment]
your CPU apparently has:
```
    * FLUSH_CMD MSR is available: NO
    * CPU indicates L1D flush capability: YES (L1D flush feature bit)
```

and the script currently does:
```
        _info_nol "* CPU microcode mitigates the vulnerability: "
        if [ "$cpu_flush_cmd" = 1 ] && [ "$cpuid_sgx" = 1 ]; then
                # no easy way to detect a fixed SGX but we know that
                # microcodes that have the FLUSH_CMD MSR also have the
                # fixed SGX (for CPUs that support it)
                pstatus green YES
        elif [ "$cpuid_sgx" = 1 ]; then
                pstatus red NO
        else
                pstatus blue N/A
        fi
```

maybe the script should use the L1D flush feature bit instead of FLUSH_CMD_MSR to determine whether CVE-2018-3615 is mitigated?

**Comment by eris23** [comment]
Well...as of today it still says:

CVE-2018-3615 aka 'Foreshadow (SGX), L1 terminal fault'
* CPU microcode mitigates the vulnerability:  NO 
> STATUS:  VULNERABLE  (your CPU supports SGX and the microcode is not up to date)

cat /proc/cpuinfo shows:

processor       : 0
vendor_id       : GenuineIntel
cpu family      : 6
model           : 142
model name      : Intel(R) Core(TM) i7-7600U CPU @ 2.80GHz
stepping        : 9
microcode       : 0xb4
...
bugs            : cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds

uname -a shows:

Linux Tablet 5.2.0-8-lowlatency #9-Ubuntu SMP PREEMPT Mon Jul 8 13:49:06 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

**Comment by Katzeilla** [comment]
Are you root?

As for me, run this scrpt will root privileges will fix this issue. :)

**Comment by eris23** [comment]
Nsame problem.  Run as root:

CVE-2018-3615 aka 'Foreshadow (SGX), L1 terminal fault'
* CPU microcode mitigates the vulnerability:  NO
> STATUS:  VULNERABLE  (your CPU supports SGX and the microcode is not up to date)

**Comment by gaul** [comment]
On my system this check fails due to `check_CVE_2018_3615` relying on the value of `cpu_flush_cmd` which is not populated due to:

```
wrmsr: pwrite: Operation not permitted
```

[StackOverflow suggests](https://stackoverflow.com/a/50474653/2800111) that disabling SecureBoot may work around this.  Is there any other way to check for this?

**Comment by Ancillius** [comment]
Can confirm on Debian Sid:
spectre-meltdown-checker:
  Installed:           0.42-2

CVE-2018-3615 aka 'Foreshadow (SGX), L1 terminal fault'
* CPU microcode mitigates the vulnerability:  NO 
> STATUS:  VULNERABLE  (your CPU supports SGX and the microcode is not up to date)

Kernel is Linux 5.3.0-2-amd64 #1 SMP Debian 5.3.9-2 (2019-11-12) x86_64
CPU is Intel(R) Core(TM) i5-8600K CPU @ 3.60GHz
  * CPU microcode is the latest known available version:  YES  (latest version is 0xb4 dated 2019/04/01 according to builtin MCExtractor DB v111 - 2019/05/18)

**Comment by martywd** [comment]
Yup, this has been an issue for me with all the 5.3.x kernels as well, but NOT with 5.2.x or 4.15.x kernels.

Examples, all from the same machine running an ubuntu 18.04 variant w/ intel-microcode 3.20191112-0ubuntu0.18.04.2

Booted into 5.3.11-050311-generic I get the 'CVE-2018-3615:KO' (VULNERABLE)
https://termbin.com/yulrv

Booted into 5.2.21-050221-generic -- 'CVE-2018-3615:**OK**' (NOT VULNERABLE)
https://termbin.com/2nbe

And finally, also on the same machine, booted into kernel 4.15.0-70-generic -- 'CVE-2018-3615:**OK**' (NOT VULNERABLE)
https://termbin.com/gvofi

.

**Comment by speed47** [comment]
@martywd 's issue root cause was different and has been fixed in #316 .
For the others: can you try the latest `master` branch? If the result doesn't seem right, please post the output of the script in very verbose mode using `-v -v`), thanks!

**Comment by Ancillius** [comment]
The current master seems to fix it:

> SUMMARY: CVE-2017-5753:OK CVE-2017-5715:OK CVE-2017-5754:OK CVE-2018-3640:OK CVE-2018-3639:OK CVE-2018-3615:OK CVE-2018-3620:OK CVE-2018-3646:OK CVE-2018-12126:OK CVE-2018-12130:OK CVE-2018-12127:OK CVE-2019-11091:OK CVE-2019-11135:OK

(no changes to the kernel/microcode)

Would be a pleasure to get this to debian sid asap (dunno whos the maintainer there).

**Comment by ilikenwf** [comment]
I also am seeing kernel lockdown mode set to max breaking this because it prevents MSR writing.

**Comment by gaul** [comment]
This still fails for me with 6e799e8b013c6543c5d1fef3f7d69ce172a9ff52 and Linux 5.3.12-300.fc31.x86_64:

`SUMMARY: CVE-2017-5753:OK CVE-2017-5715:OK CVE-2017-5754:OK CVE-2018-3640:OK CVE-2018-3639:OK CVE-2018-3615:KO CVE-2018-3620:OK CVE-2018-3646:OK CVE-2018-12126:OK CVE-2018-12130:OK CVE-2018-12127:OK CVE-2019-11091:OK CVE-2019-11135:OK CVE-2018-12207:OK`

**Comment by speed47** [comment]
Do any of you @ilikenwf  or @gaul  (or both!) run the script with `-v -v` and post the output? This would help pinpointing the problem. I'd like to fix this before releasing the next version. Thanks!

**Comment by ilikenwf** [comment]
I'm not sure it's the same issue - but the lockdown feature, if enabled,
has multiple options...

In the highest setting, the MSRs can't be unlocked via LSM, in the next
highest, in theory, a proper LSM policy manager (apparmor etc) could
allow access for spectre-meltdown-checker.

It needs some reworking, I think, to allow whitelisting of specific
registers. I have a thread going on the linux-security-module mailing
list right now about it.

**Comment by gaul** [comment]
With 0cd7e1164f1ebcbcc13484fe1b1218f1154ecbb2 spectre-meltdown-checker reports:

```
  * L1 data cache invalidation
    * FLUSH_CMD MSR is available:  UNKNOWN  (your kernel is locked down (Fedora/Red Hat), please reboot without secure boot and retry)
...
> SUMMARY: CVE-2017-5753:OK CVE-2017-5715:OK CVE-2017-5754:OK CVE-2018-3640:OK CVE-2018-3639:OK CVE-2018-3615:OK CVE-2018-3620:OK CVE-2018-3646:OK CVE-2018-12126:OK CVE-2018-12130:OK CVE-2018-12127:OK CVE-2019-11091:OK CVE-2019-11135:OK CVE-2018-12207:OK
```

**Comment by speed47** [comment]
When lockdown is detected (either by the Red Hat patch, or the more recent one that made it to vanilla in 5.4+), the scripts only relies on CPUID and ignores the fact that it can't test the write-only MSRs.
This should fix the issue indeed.

**Comment by speed47** [comment]
This issue had a lot of reporters, can it be considered fixed with latest release?

**Comment by Photon89** [comment]
Just ran into it. Maybe add something like "The analysis might be inaccurate due to permission problems, please run as root!" as last output line, if there are permission issues?

**Comment by kjkent** [comment]
Unfortunately it seems this issue is still present, or is present again (at least for me). 

Running `./spex.sh --allow-msr-write --explain` as root returns the unexpected result on my i5-8350u with latest microcode applied.

**Comment by speed47** [comment]
Sorry for the delay on this one, I've implemented a two-folds fix in the test version available [here](https://raw.githubusercontent.com/speed47/spectre-meltdown-checker/refs/heads/test-build/spectre-meltdown-checker.sh):

- Add probing of `/sys/kernel/security/lockdown` to proactively detect lockdown (kernels 5.4+) in addition to the dmesg grep
- No longer require the FLUSH_CMD MSR probing to deem this CVE as mitigated, as per Intel guidance the CPUID is enough, and when it's present, the MSR is always present too.

**Comment by speed47** [comment]
Released https://github.com/speed47/spectre-meltdown-checker/releases
